fix(core): Error on invalid APP_ID

JeanMeche opened the pull request:
Unsanitized appIds could be responsible to generating broken CSS selectors. (eg `:` is an example for a character that breaks a selector by being a separator for pseudo-selectors.)

fixes angular#63251

JeanMeche commented:
We'll need a TGP on this to see if this is a breaking change.

thePunderWoman commented:
[TGP](https://test.corp.google.com/OCL:807770865:BASE:807678934:1758047442436:c201075)

AndrewKushnir reviewed (commented):
Shared some feedback with @JeanMeche via chat. Key points:

- The check should happen sooner (e.g. at the time we create an application injector, using the `ENVIRONMENT_INITIALIZER` callback).
- We should throw an error if the `APP_ID` value is "unsafe", so that we don't need to keep sanitizing it in the codebase at the time when we need to use it (easy to forget adding sanitization at that point).

JeanMeche committed:
fix(core): Error on invalid APP_ID

An invalid APP_ID could be responsible to generating broken CSS selectors. (eg `:` is an example for a character that breaks a selector by being a separator for pseudo-selectors.)
We now throw an error if the provided value is not alphanumerical

AndrewKushnir reviewed (commented):
Looks great, thanks @JeanMeche!

A couple of comments:

- It'd be great to add some tests to verify this behavior
- We should consider adding a similar check to the NgModule-based bootstrap logic as well

thePunderWoman commented:
@JeanMeche This has conflicts that need to be resolved.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

jelbourn commented on packages/core/src/application/application_tokens.ts:
```diff
@@ -47,6 +49,25 @@ export const APP_ID = new InjectionToken<string>(ngDevMode ? 'AppId' : '', {
 const DEFAULT_APP_ID = 'ng';
 
+/** Initializer check for the validity of the APP_ID */
+export const appIdSanityCheckInitializer: StaticProvider = {
```
Can we only add this in dev mode?

jelbourn replied:
FYI we generally want to avoid the term "sanity" for things like this these days, so maybe something like `validAppIdInitializer`

thePunderWoman commented:
There are a few failures, but I'm pretty sure they're entirely unrelated to this based on the traces. I think this is safe to merge.

thePunderWoman reviewed (approved):
LGTM
